File: langs/rust/tests/conformance.rs
//! Teste de conformidade: roda os vetores sintéticos em
//! `conformance/vectors.json` e compara com `expected_md5` da referência
//! Python.
//!
//! Cada vetor tem um campo opcional `expect`:
//! - ausente ou `"hash"` → POSITIVO: `hash_tiss` deve `Ok` e bater com
//!   `expected_md5`.
//! - `"error"` → NEGATIVO: `hash_tiss` deve `Err` (e `expected_md5` é
//!   `null`).
//!
//! Localização dos artefatos: `../../conformance/` (relativo a este arquivo,
//! que vive em `langs/rust/tests/`). Resolução é feita via
//! `CARGO_MANIFEST_DIR` para garantir robustez independente do CWD do
//! runner.

use serde::Deserialize;
use std::path::{Path, PathBuf};

#[derive(Deserialize)]
struct Manifest {
    vectors: Vec<Vector>,
}

#[derive(Deserialize)]
struct Vector {
    id: String,
    input: String,
    /// `null` em vetores negativos (`expect == "error"`).
    #[serde(default)]
    expected_md5: Option<String>,
    /// Ausente/`"hash"` = positivo; `"error"` = negativo.
    #[serde(default)]
    expect: Option<String>,
    #[serde(default)]
    #[allow(dead_code)]
    desc: String,
}

impl Vector {
    /// `true` se o vetor é negativo (port deve rejeitar com `Err`).
    fn is_error(&self) -> bool {
        self.expect.as_deref() == Some("error")
    }
}

fn conformance_dir() -> PathBuf {
    // CARGO_MANIFEST_DIR = .../langs/rust → .../conformance
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest
        .parent() // langs/
        .and_then(|p| p.parent()) // raiz do repo
        .map(|p| p.join("conformance"))
        .expect("não foi possível resolver conformance/ a partir de CARGO_MANIFEST_DIR")
}

fn load_manifest() -> Manifest {
    let path = conformance_dir().join("vectors.json");
    let raw =
        std::fs::read(&path).unwrap_or_else(|e| panic!("falha ao ler {}: {e}", path.display()));
    serde_json::from_slice(&raw)
        .unwrap_or_else(|e| panic!("vectors.json mal-formado em {}: {e}", path.display()))
}

#[test]
fn todos_vetores_passam() {
    let manifest = load_manifest();
    let dir = conformance_dir();
    let mut fails: Vec<String> = Vec::new();

    for v in &manifest.vectors {
        let input_path = dir.join(&v.input);
        let raw = match std::fs::read(&input_path) {
            Ok(b) => b,
            Err(e) => {
                fails.push(format!("[IO] {}: {} ({e})", v.id, input_path.display()));
                continue;
            }
        };
        let result = tiss_hash::hash_tiss(&raw);

        if v.is_error() {
            // NEGATIVO: deve rejeitar.
            match result {
                Err(e) => println!("[OK-ERR] {:33} rejeitado: {e}", v.id),
                Ok(got) => fails.push(format!(
                    "[SHOULD-ERR] {:29} esperado Err, obteve Ok({got})",
                    v.id
                )),
            }
            continue;
        }

        // POSITIVO: deve casar com expected_md5.
        let expected = match v.expected_md5.as_deref() {
            Some(e) => e,
            None => {
                fails.push(format!(
                    "[BAD-VEC] {:32} vetor positivo sem expected_md5",
                    v.id
                ));
                continue;
            }
        };
        match result {
            Ok(got) if got == expected => {
                println!("[OK]   {:35} {got}", v.id);
            }
            Ok(got) => {
                fails.push(format!("[DIFF] {:35} got={got} expected={expected}", v.id));
            }
            Err(e) => {
                fails.push(format!("[ERR]  {:35} {e}", v.id));
            }
        }
    }

    assert!(
        fails.is_empty(),
        "{}/{} vetores falharam:\n{}",
        fails.len(),
        manifest.vectors.len(),
        fails.join("\n")
    );
}

#[test]
fn vetor_minimal_isolado() {
    // Sanidade: garante que a função básica funciona mesmo se o JSON for
    // alterado por engano. Hash colado fixo da referência.
    let dir = conformance_dir();
    let raw = std::fs::read(dir.join("inputs/syn_minimal.xml")).unwrap();
    let got = tiss_hash::hash_tiss(&raw).unwrap();
    assert_eq!(got, "3aa0c578c95cdb861a125f480a8a4de5");
}

#[test]
fn hash_tiss_file_funciona() {
    // hash_tiss_file deve produzir o mesmo resultado de hash_tiss(read).
    let dir = conformance_dir();
    let path = dir.join("inputs/syn_acento.xml");
    let h1 = tiss_hash::hash_tiss_file(&path).unwrap();
    let h2 = tiss_hash::hash_tiss(&std::fs::read(&path).unwrap()).unwrap();
    assert_eq!(h1, h2);
    assert_eq!(h1, "a20afc9a89aadaa2179d03d225337662");
}

File: langs/wasm/test/conformance.mjs
#!/usr/bin/env node
// Harness de conformidade do port WASM (carregado no Node).
//
// Carrega o módulo WASM gerado por wasm-bindgen (pkg/node, target nodejs),
// roda os 20 vetores sintéticos de ../../conformance/vectors.json e, se
// disponíveis, os 3 goldens reais privados.
//
// Semântica dos vetores (igual aos outros ports):
//   - `expect` ausente ou "hash" => POSITIVO: hashTiss deve retornar e o hash
//     deve casar com expected_md5.
//   - `expect` === "error"       => NEGATIVO: hashTiss deve lançar exceção.
//
// PRIVACIDADE (LGPD): os goldens reais contêm PII. Este harness só imprime
// PASS/FAIL por arquivo: NUNCA o hash, NUNCA o conteúdo do XML, NUNCA a
// versão TISS ou a operadora. O bg.wasm e os bindings rodam tudo em memória.

import { readFileSync, existsSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join, resolve } from 'node:path';
import { createRequire } from 'node:module';

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, '..', '..', '..'); // langs/wasm/test -> raiz
const conformanceDir = join(repoRoot, 'conformance');

// O binding target=nodejs é CommonJS; carrega via createRequire.
const require = createRequire(import.meta.url);
const wasmModulePath = join(__dirname, '..', 'pkg', 'node', 'tiss_hash_wasm.js');
const { hashTiss, tissNamespace } = require(wasmModulePath);

let pass = 0;
let fail = 0;
const failures = [];

function ok(msg) {
  pass += 1;
  console.log(`[OK]   ${msg}`);
}

function bad(msg) {
  fail += 1;
  failures.push(msg);
  console.log(`[FAIL] ${msg}`);
}

// --- Sanidade da API ------------------------------------------------------

const NS = 'http://www.ans.gov.br/padroes/tiss/schemas';
if (tissNamespace() === NS) {
  ok(`tissNamespace() = namespace TISS canônico`);
} else {
  bad(`tissNamespace() divergente: ${tissNamespace()}`);
}

// --- Vetores sintéticos ---------------------------------------------------

const manifest = JSON.parse(
  readFileSync(join(conformanceDir, 'vectors.json'), 'utf8'),
);

console.log(`\n== Vetores sintéticos (${manifest.vectors.length}) ==`);
for (const v of manifest.vectors) {
  const inputPath = join(conformanceDir, v.input);
  const isError = v.expect === 'error';
  let raw;
  try {
    raw = readFileSync(inputPath); // Buffer (subtype de Uint8Array)
  } catch (e) {
    bad(`${v.id} :: não foi possível ler ${v.input}: ${e.message}`);
    continue;
  }

  // hashTiss aceita Uint8Array; Buffer já é Uint8Array no Node, mas
  // normalizamos para deixar explícito o contrato (e exercitar Uint8Array).
  const bytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);

  if (isError) {
    // NEGATIVO: deve lançar.
    try {
      const got = hashTiss(bytes);
      bad(`${v.id} :: esperado erro, obteve hash (len=${got.length})`);
    } catch {
      ok(`${v.id} :: rejeitado como esperado`);
    }
    continue;
  }

  // POSITIVO: deve casar com expected_md5.
  if (typeof v.expected_md5 !== 'string') {
    bad(`${v.id} :: vetor positivo sem expected_md5 no manifesto`);
    continue;
  }
  try {
    const got = hashTiss(bytes);
    if (got === v.expected_md5) {
      ok(`${v.id} :: ${got}`);
    } else {
      bad(`${v.id} :: got=${got} expected=${v.expected_md5}`);
    }
  } catch (e) {
    bad(`${v.id} :: erro inesperado: ${e.message}`);
  }
}

// --- Goldens reais privados (PASS/FAIL apenas) ----------------------------
//
// Localização: env TISS_PRIVATE_XMLS, senão o path default ao lado do repo.
// Nunca commitados (LGPD). Se ausentes, é SKIP (não falha a suite).

const privateDir =
  process.env.TISS_PRIVATE_XMLS ||
  resolve(repoRoot, '..', '_private_tiss_real_xmls');
const expectedPath = join(privateDir, 'expected_hashes.json');

console.log('\n== Goldens reais privados (PASS/FAIL apenas; sem PII) ==');
if (!existsSync(expectedPath)) {
  console.log(
    `[SKIP] goldens privados ausentes (${privateDir}). ` +
      `Defina TISS_PRIVATE_XMLS para validar. Não afeta os vetores sintéticos.`,
  );
} else {
  const expected = JSON.parse(readFileSync(expectedPath, 'utf8'));
  let goldenPass = 0;
  let goldenTotal = 0;
  for (const [fileName, expectedHash] of Object.entries(expected)) {
    goldenTotal += 1;
    const xmlPath = join(privateDir, fileName);
    if (!existsSync(xmlPath)) {
      bad(`golden ${goldenTotal}/${Object.keys(expected).length} :: arquivo ausente`);
      continue;
    }
    try {
      const raw = readFileSync(xmlPath);
      const bytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
      const got = hashTiss(bytes);
      // Compara em memória; NUNCA imprime got nem expectedHash.
      if (got === expectedHash) {
        goldenPass += 1;
        ok(`golden ${goldenPass}/${Object.keys(expected).length} :: PASS`);
      } else {
        bad(`golden ${goldenTotal} :: FAIL (hash divergente, não exibido)`);
      }
    } catch (e) {
      // A mensagem do core não contém PII, mas por garantia não a exibimos.
      bad(`golden ${goldenTotal} :: FAIL (exceção inesperada)`);
    }
  }
  console.log(`Goldens: ${goldenPass}/${goldenTotal} PASS`);
}

// --- Resumo ---------------------------------------------------------------

console.log(`\n==================================================`);
console.log(`Resultado: ${pass} PASS, ${fail} FAIL`);
if (fail > 0) {
  console.error('\nFalhas:');
  for (const f of failures) console.error(`  - ${f}`);
  process.exit(1);
}
console.log('Tudo verde.');
process.exit(0);
